File: tyk/middleware/transform-response.js
// Transform Response Middleware
// Runs after the upstream backend responds (response hook)
var transformResponse = new TykJS.TykMiddleware.NewMiddleware({});

transformResponse.NewProcessRequest(function(request, session, config) {
    // Response body is available as a base64-encoded string in request.Body
    var rawBody = b64dec(request.Body);

    try {
        var body = JSON.parse(rawBody);

        // 1. Rename "message" to "status"
        if (body.message !== undefined) {
            body.status = body.message;
            delete body.message;
        }

        // 2. Rename "channelReceived" to "channel"
        if (body.channelReceived !== undefined) {
            body.channel = body.channelReceived;
            delete body.channelReceived;
        }

        // 3. Rename "prefixReceived" to "prefix"
        if (body.prefixReceived !== undefined) {
            body.prefix = body.prefixReceived;
            delete body.prefixReceived;
        }

        // 4. Add metadata
        body.transformed = true;

        // Re-encode body back to base64
        request.Body = b64enc(JSON.stringify(body));

        log("transform-response: body transformed successfully");

    } catch (e) {
        log("transform-response: failed to parse body - " + e);
    }

    return transformResponse.ReturnData(request, {});
});

log("Transform-response middleware loaded");

File: tyk/go-plugins/transform_response.go
package main

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
)

// BackendResponse matches the shape of dummy-backend's /hello response.
type BackendResponse struct {
	Message         string `json:"message"`
	ChannelReceived string `json:"channelReceived"`
	PrefixReceived  string `json:"prefixReceived"`
}

// TransformedResponse is the outbound shape after transformation.
type TransformedResponse struct {
	Status      string `json:"status"`
	Channel     string `json:"channel"`
	Prefix      string `json:"prefix"`
	Transformed bool   `json:"transformed"`
}

// TransformResponse is the response Go plugin equivalent of the transform_response
// Go template. It renames fields and adds transformed:true.
func TransformResponse(rw http.ResponseWriter, res *http.Response, req *http.Request) {
	if res == nil {
		return
	}

	body, err := io.ReadAll(res.Body)
	if err != nil {
		logger.WithError(err).Error("TransformResponse: failed to read response body")
		return
	}
	res.Body.Close()

	var upstream BackendResponse
	if err := json.Unmarshal(body, &upstream); err != nil {
		logger.WithError(err).Error("TransformResponse: failed to parse response body")
		// Pass through unchanged if we can't parse
		res.Body = io.NopCloser(bytes.NewBuffer(body))
		return
	}

	transformed := TransformedResponse{
		Status:      upstream.Message,
		Channel:     upstream.ChannelReceived,
		Prefix:      upstream.PrefixReceived,
		Transformed: true,
	}

	newBody, err := json.Marshal(transformed)
	if err != nil {
		logger.WithError(err).Error("TransformResponse: failed to marshal transformed body")
		res.Body = io.NopCloser(bytes.NewBuffer(body))
		return
	}

	res.Body = io.NopCloser(bytes.NewBuffer(newBody))
	res.Header.Set("Content-Type", "application/json")
	res.ContentLength = int64(len(newBody))

	logger.Info("TransformResponse: response body transformed successfully")
}
